### PR TITLE
Check placeholder records and handle validation

### LINKED ISSUE #78: Linter: accept "no authorized advertising systems" placerholder as valid.
**Is your enhancement related to a problem? Please describe.**

The ads.txt specification recommends sites without authorised advertising sellers use the following placeholder:
```
placeholder.example.com, placeholder, DIRECT, placeholder
```

Currently the built in linting of the ads.txt file highlights that as an error.

Screen Shot 2022-01-14 at 3 17 13 pm

**Describe the solution you'd like**

For sites using the placeholder text above as the only text in the file, accept the content as valid but warn it indicates there are no authorized resellers.

**Designs**

Hacked together mock-up that would need work.

Screen Shot 2022-01-14 at 3 20 54 pm

Note: color code changed to yellow as warning rather than an error.

**Describe alternatives you've considered**

I considered suppressing the error message altogether but decided a warning of no authorized sellers would be more appropriate given the most likely ads.txt use case.

### PR DESCRIPTION
**Pull request by ankitrox**
### Description of the Change

Closes #78 

### How to test the Change

1. Add the following record in ads.txt editor and save changes. You shall receive the warning "Your ads.txt indicates no authorized advertising sellers."
```
placeholder.example.com, placeholder, DIRECT, placeholder
```

2. Add following record in ads.txt editor. You will get the error "Line 2: Ads.txt contains placeholder record with another records."
```
google.com, 4254, DIRECT
placeholder.example.com, placeholder, DIRECT, placeholder
```

### Changelog Entry

> Added - Placeholder record can be added when no authorized sellers or buyers. 

### Credits

Props @peterwilsoncc

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests pass.

### CODE REVIEW

**Review by peterwilsoncc — COMMENTED** [review]
@ankitrox Thanks for the pull request.

Rather than use a static, would it be possible to check if both placeholder and valid records exist within the save function?

https://github.com/10up/ads-txt/blob/a00bdce22ec238acc0d1d599c448bb62af53d2db/inc/save.php#L34-L46

Statics can be difficult to check and while testing the changes with the `test_validate_line` function, I discovered the static was affecting other tests.

**Comment by jeffpaul** [comment]
@ankitrox any chance you're able to handle the code review item above?

**Comment by ankitrox** [comment]
@peterwilsoncc Thanks for the review :) 

I wanted those couple of variables to be persistent across multiple function calls, but I agree that static may not be the idea ones as far as unit tests are concerned.

I have changed it to use options table in order for them to be persistent. This will also be handy for unit tests.

File: inc/save.php [hunk 1]
```diff
@@ -7,6 +7,16 @@
...
+// Option name for a flag to tell if ads.txt contains placeholder record.
+if ( ! defined( 'ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME' ) ) {
+	define( 'ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME', 'ads_txt_has_placeholder_record' );
+}
+
+// Option name for a flag to store total number of record lines (temporary).
+if ( ! defined( 'ADS_TXT_RECORD_LINES_OPTION_NAME' ) ) {
+	define( 'ADS_TXT_RECORD_LINES_OPTION_NAME', 'ads_txt_record_lines' );
+}
+
```
Comment (peterwilsoncc) [inline inc/save.php]:
```suggestion
```

File: inc/save.php [hunk 2]
```diff
@@ -28,6 +38,7 @@ function save() {
 	$lines     = preg_split( '/\r\n|\r|\n/', $_post['adstxt'] );
 	$sanitized = array();
 	$errors    = array();
+	$warnings  = array();
 	$response  = array();
```
Comment (peterwilsoncc) [inline inc/save.php]:
```suggestion
	$response  = array();
	$has_placeholder_record = false;
```

File: inc/save.php [hunk 3]
```diff
@@ -38,8 +49,16 @@ function save() {
 		if ( ! empty( $result['errors'] ) ) {
```
Comment (peterwilsoncc) [inline inc/save.php:40]: Above, change the `validate_line()` call to

```php
$result      = validate_line( $line, $line_number, $has_placeholder_record );
```

(Sorry, I can't suggest as it's outside the diff)

File: inc/save.php [hunk 4]
```diff
@@ -38,8 +49,16 @@ function save() {
...
 			$errors = array_merge( $errors, $result['errors'] );
 		}
+
+		if ( ! empty( $result['warnings'] ) ) {
+			$warnings = array_merge( $warnings, $result['warnings'] );
+		}
```
Comment (peterwilsoncc) [inline inc/save.php:46]:
```suggestion
		}

		if ( ! empty( $result['is_placeholder_record'] ) ) {
			$has_placeholder_record = true;
		}
```

File: inc/save.php [hunk 5]
```diff
@@ -38,8 +49,16 @@ function save() {
...
 	}
 
+	// Delete the temporary options.
+	delete_option( ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME );
+	delete_option( ADS_TXT_RECORD_LINES_OPTION_NAME );
+
```
Comment (peterwilsoncc) [inline inc/save.php]:
```suggestion
```

File: inc/save.php [hunk 6]
```diff
@@ -96,8 +116,12 @@ function save() {
  * }
  */
 function validate_line( $line, $line_number ) {
```
Comment (peterwilsoncc) [inline inc/save.php]:
```suggestion
function validate_line( $line, $line_number, $has_placeholder_record = false ) {
```

Will need a docblock changes but I can't add it as it's outside the diff.

File: inc/save.php [hunk 7]
```diff
@@ -96,8 +116,12 @@ function save() {
...
+	$has_placeholder_record = get_option( ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME, false );
+	$record_lines           = get_option( ADS_TXT_RECORD_LINES_OPTION_NAME, 0 );
```
Comment (peterwilsoncc) [inline inc/save.php]:
```suggestion
	static $record_lines           = 0;
	$is_placeholder_record = false;
```

File: inc/save.php [hunk 8]
```diff
@@ -141,37 +165,69 @@ function validate_line( $line, $line_number ) {
 		$fields = explode( ',', $record );
 
 		if ( 3 <= count( $fields ) ) {
-			$exchange     = trim( $fields[0] );
-			$pub_id       = trim( $fields[1] );
-			$account_type = trim( $fields[2] );
+			++$record_lines;
+			update_option( ADS_TXT_RECORD_LINES_OPTION_NAME, $record_lines ); // Increment record lines.
```
Comment (peterwilsoncc) [inline inc/save.php]:
```suggestion
```

File: inc/save.php [hunk 9]
```diff
@@ -141,37 +165,69 @@ function validate_line( $line, $line_number ) {
...
+			$is_placeholder_record = is_placeholder_record( $exchange, $pub_id, $account_type, $fields[3] ? trim( $fields[3] ) : null );
 
-			if ( ! preg_match( $domain_regex, $exchange ) ) {
-				$errors[] = array(
-					'line'  => $line_number,
-					'type'  => 'invalid_exchange',
-					'value' => $exchange,
+			// If the file contains placeholder record and no placeholder was already present, set variable.
+			if ( $is_placeholder_record && ! $has_placeholder_record ) {
+				update_option( ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME, true, false );
```
Comment (peterwilsoncc) [inline inc/save.php]:
```suggestion
```

File: inc/save.php [hunk 10]
```diff
@@ -192,6 +248,7 @@ function validate_line( $line, $line_number ) {
 	return array(
 		'sanitized' => $sanitized,
 		'errors'    => $errors,
+		'warnings'  => $warnings,
```
Comment (peterwilsoncc) [inline inc/save.php]:
```suggestion
		'warnings'  => $warnings,
		'is_placeholder_record' => $is_placeholder_record,
```

**Review by peterwilsoncc — CHANGES_REQUESTED** [review]
I've added quite a few suggestions inline to avoid thrashing the database with option updates. 

To record the line number, I've switch from an option to a static variable. (This may need further thought)

To record the presence of a placeholder, I've added a value to the returned array in `validate_line()`.

While testing, I noticed that the order of lines affects the warnings shown. If I place a valid record above the placeholder record I am shown the warning "Your ads.txt indicates no authorized advertising sellers."

Screen Shot 2023-07-03 at 10 43 52 am

If I place the two lines in the reverse order, I also get a message that "Ads.txt contains placeholder record with another records."

Screen Shot 2023-07-03 at 10 46 26 am

**Comment by peterwilsoncc** [comment]
I just tested this and am still seeing the no authorized resellers warning when the file contains both a placeholder and a valid record. 

I did some testing to see if I could come up with an approach that worked but comment lines ended up suppressing the error if the file contained placeholders and comments (which is just as bad, if not worse).

My thoughts are:

* rename `has_placeholder_records` to `has_only_placeholder_records`, in the validate_file function set it to `null`
* if we find a valid line, set the renamed variable to `false`
* if we find a placeholder and the existing value is `null`, set it to true.
* show the warning if the value is either null or true

We'll need to account for:
* comments: no affect on the value
* empty lines: no affect on the value
* invalid records: no affect on the value
* valid record
* placeholders

**Comment by ankitrox** [comment]
Thank you @peterwilsoncc 

I have added the fix as per Peter's suggestion and it seems to be working fine. Also, added the unit test for the placeholder record.

CC: @jeffpaul

**Review by peterwilsoncc — APPROVED** [review]
This looks good to me and tests well. Thank you so much for your patience as we got this working.

I pushed a couple of minor changes:
* 811fabca5eb4910cb4e24288701e4e0711af6235 to make sure the optional tag value was set
* f1a56e501b201458e92e5565a633f5c56719b104 to remove some sniff suppression that were not needed.

**Pull request merged**